grpc-protocol/src/main/java/com/linecorp/armeria/client/grpc/protocol/UnaryGrpcClient.java

```diff
@@ -239,33 +289,33 @@ public void onNext(DeframedMessage message) {
                         return;
                     }
                     if (processedMessages > 0) {
-                        onError(new ArmeriaStatusException(StatusCodes.INTERNAL,
-                                                           "received more than one data message; " +
-                                                           "UnaryGrpcClient does not support streaming."));
+                        cancel(new ArmeriaStatusException(StatusCodes.INTERNAL,
+                                                          "received more than one data message; " +
+                                                          "UnaryGrpcClient does not support streaming."));
                         return;
                     }
                     final ByteBuf buf = message.buf();
                     // Compression not supported.
                     assert buf != null;
                     content = HttpData.wrap(buf);
+                    buf.retain();
                     processedMessages++;
                 }
 
-                @Override
-                public void onError(Throwable t) {
-                    content.close();
-                    responseFuture.completeExceptionally(t);
-                }
-
-                @Override
-                public void onComplete() {
-                    if (responseFuture.isDone()) {
+                private void cancel(Throwable t) {
+                    if (terminated) {
                         return;
                     }
-                    if (trailers == null) {
-                        trailers = msg.trailers();
+                    terminated = true;
+                    if (subscription != null) {
+                        subscription.cancel();
                     }
```

Comment:
`subscription` should not be null here. `onNext()`, `onError()` and `onComplete()` MUST be called after `onSubscribe()`. 
Isn't it better to add `assert` than `if` condition?

Reply:
Not sure, let's discuss. My reading of the reactive streams spec is that Subscriber that follows the spec doesn't throw exceptions, except when any provided parameter is null:

```
Calling onSubscribe, onNext, onError or onComplete MUST return normally except when any provided parameter is null in which case it MUST throw a java.lang.NullPointerException to the caller, for all other situations the only legal way for a Subscriber to signal failure is by cancelling its Subscription. 
```

`assert` will throw, and I'm not sure if that's okay to do since that may cause issues to cascade and make it hard to debug. If you're concerned about this issue being handled silently - how about logging an error in case subscription is null here?

Reply:
It is agreeable. However `assert` has a different meaning compared to other exceptions. 
1. It will be removed from our user's code if `-ea` option is not specified.
  - This would be good for micro optimizations.
  - We don't want to execute `if` and `else` condition if it should be always true.
2. `assert` is commonly used to help IDE and devs statically analyze a stateful code.
3. We assume that `assert` MUST not raise `AssertionError`.

Reply:
> It is agreeable. However `assert` has a different meaning compared to other exceptions.
> 
> 1. It will be removed from our user's code if `-ea` option is not specified.
    
If `assert` is removed from the user's code and `subscription` is `null` - it'll unexpectedly throw NullPointerException for `subscription.cancel` which would violate the spec - `subscription` is not a is not a provided param for `onNext`. By using `assert` we will intentionally make this subscriber implementation faulty, and with various flavors of faultiness based on the VM options :)

>    * This would be good for micro optimizations.
>    * We don't want to execute `if` and `else` condition if it should be always true.

Modern hardware usually doesn't benefit from micro optimizations like this, conditions that always true are usually cheap to execute thanks to https://en.wikipedia.org/wiki/Branch_predictor.

Micro optimizations are usually only good for micro benchmarking. In real world scenarios code is executed in the conditions that are hard to reproduce in the micro benchmarks, and is limited by other factors. If this method doesn't take a significant % of the CPU flame graph in the production - it's not worth optimizing.

Unless I'm missing something about Java world and you have a way to demonstrate that using if condition here would have a non-negligible impact on Armeria's performance.

> 2. `assert` is commonly used to help IDE and devs statically analyze a stateful code.
> 3. We assume that `assert` MUST not raise `AssertionError`.

True, `assert` is used to document the assumptions in the code, though in this particular case it'd lead to violating the spec since it doesn't mention that `AssertionError` is allowed to be raised, and it MAY be raised if e.g. bug is introduced in the publisher in the future. If that happens - we probably don't want servers in production to be broken by the AsserionError or NullPointerException raised unexpectedly from the subscriber. You can make an argument that given faulty producer it's okay to have faulty subscriber, but I think we should prefer to follow the spec:

1. It'll make things easier for developers familiar with the spec to contribute
2. Less edge cases to consider on publisher side
3. Higher chances that misbehavior of one part of the code won't cause a cascading failure elsewhere, e.g. because publisher wasn't expecting such behavior.

Reply:
>  though in this particular case it'd lead to violating the spec since it doesn't mention that AssertionError is allowed to be raised, 

It is an important specification not to throw any exception from `Subscriber`. 👍 
However, it might be meaningful when a `Publisher` follows the Reactive Streams specifications.
A `Publisher` should pass a Subscription via `onSubscription(Subsription)` first and other methods have to be invoked. A `Subscriber` is a small state machine. The correct state could be preserved only when a `Publisher` correctly calls the APIs in order.

If the state is broken by the `Publisher`, can we throw `AssertionError` or `IllegalStateException`?
Furthermore, in this case, the `Publisher`s are Armeria's `StreamMessage` that we manage.
Do you think `assert` is still harmful?

> Modern hardware usually doesn't benefit from micro optimizations like this,
Micro optimizations seem overkill. It does not really matter.

Reply:
> > though in this particular case it'd lead to violating the spec since it doesn't mention that AssertionError is allowed to be raised,
> 
> It is an important specification not to throw any exception from `Subscriber`. 👍
> However, it might be meaningful when a `Publisher` follows the Reactive Streams specifications.
> A `Publisher` should pass a Subscription via `onSubscription(Subsription)` first and other methods have to be invoked. A `Subscriber` is a small state machine. The correct state could be preserved only when a `Publisher` correctly calls the APIs in order.
> If the state is broken by the `Publisher`, can we throw `AssertionError` or `IllegalStateException`?

Looking at the provided examples - they don't thrown in such cases. E.g. https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.3/examples/src/main/java/org/reactivestreams/example/unicast/SyncSubscriber.java#L56 - creates `IllegalStateException` and prints stack trace (but doesn't throw).

> Furthermore, in this case, the `Publisher`s are Armeria's `StreamMessage` that we manage.
> Do you think `assert` is still harmful?

I still do. "return normally" requirement (2.13 in the spec) allows for a straightforward publisher - subscriber interaction - publisher may call `onNext` like this:

```
Stream<T> stream = streamSupplier.get();
stream.forEach(subscriber::onNext);
```

and can reasonably expect that subscriber will normally process or discard all the messages provided on that stream. I we break that invariant (even if that happens due to the faulty producer that "forgot" to call `onSubscribe`) - I can imagine at least a couple of things that can go wrong and could be avoided by strictly following the spec:

1. Depending on the contract between publisher and subscriber - subscriber may be responsible for managing resources (like buffer refcnt in this PR). If `onNext` returns exceptionally - other `onNext` calls in that examples won't be called and there will be a leak. It's expected that even in case of cancellation subscriber may receive `onNext` calls with pending messages https://github.com/reactive-streams/reactive-streams-jvm#2.8 where it can properly release resources if needed.
2. If `onNext` raises - you need to carefully think about how onNext is structured, taking into account not only the regular flow control, but also a possible exception. E.g. in `DeframedMessageSubscriber.onNext` raising an exception directly in place of current `if` condition would result in `completeExceptionally` being skipped and never called. You could argue that fix is as simple as rearranging the lines, but I just don't need to worry about that because I follow the contract and make sure nothing is thrown, `completeExceptionally` is guaranteed to be called even if publisher is faulty and `subscription == null`.

I think there are 2 options:
1. If Armeria wants to conform to the reactive streams spec - it should follow the specification rules strictly. This will help with future contributions and compatibility with other libraries that follow the spec. This spec is a contract that helps developers from multiple projects write software that's compatible and bug-free. Since this model seems to be successful and quite popular in Java world - I'd assume that rules provided in the spec are there for a reason, and we should not diverge from them. If something is not permitted in the spec **explicitly** - it's not allowed. If somehow we need to bend the rules to make something work - there's a high chance that we are using that in a wrong and unexpected way.
2. If reactive streams specification is considered confusing - Armeria should use custom interfaces that are better suited for the job (note that I don't necessarily recommend that 😄 ). The problem we started with is that subscription may not be set when `onNext` is called, that's happening because subscriber is following the reactive streams approach of setting the subscription **after** the object creation, which means there's a part of object lifetime where subscription is `null` - that problem would not have existed if subscription object was provided in the subscriber constructor and verified to be not `null` before any interactions between the publisher and subscriber.

I think it's a bad idea to try to conform to the specification and at the same time occasionally break the rules based on the assumption that "this situation will never happen", because it occasionally will 😉